src/Chapter4.hs

```diff
@@ -652,9 +654,89 @@ concatList (Cons x xs) = append x (concatList xs)
 
 instance Monad (List) where
     (>>=) :: List a -> (a -> List b) -> List b
-    Empty >>= _ = Empty
     el >>= f = concatList (fmap f el)
 
+{- |
+Task 7.1 
+Convert two maybes to a single maybe of a pair:
+>>> maybePair (Just 1) (Just 2)
+Just (1,2)
+>>> maybePair (Just 1) Nothing
+Nothing
+>>> maybePair Nothing Nothing
+Nothing
+-}
+maybePair :: Maybe a -> Maybe b -> Maybe (a, b)
+maybePair (Just x1) (Just x2) = Just (x1, x2)
+maybePair _ _ = Nothing
+
+{- |
+Task 7.2 
+Find a sum of two maybes:
+-}
+maybeSum :: Maybe Int -> Maybe Int -> Maybe Int
+maybeSum (Just x1) (Just x2) = Just (x1 + x2)
+maybeSum (Just x1) Nothing = Just x1
+maybeSum Nothing (Just x2) = Just x2
+maybeSum _ _ = Nothing
+
+{- |
+Task 7.3
+Can you generalize the above function and notice the pattern? Something like:
+bind2 :: ??? -> Maybe a -> Maybe b -> Maybe c
+>>> bind2 (+) (Just 1 ) (Just 2 )
+Just 3
+>>> bind2 (+) (Just 4 ) (Just 2 )
+Just 6
+>>> bind2 (*) (Just 4 ) (Just 2 )
+Just 8
+>>> bind2 (*) (Just 4 ) (Nothing)
+Nothing
+-}
+
+bind2 :: ( a -> b -> c ) -> Maybe a -> Maybe b -> Maybe c
+bind2 f (Just x1) (Just x2) = Just (f x1 x2)
+bind2 _ _ _ = Nothing
+
+{- |
+Task 7.4
+Try implementing smartReplicate for your List using Monad instance for lists.
+@
+fmap  ::   (a -> b) -> f a -> f b
+(<*>) :: f (a -> b) -> f a -> f b
+(=<<) :: (a -> f b) -> f a -> f b
+(>>=) :: f a -> (a -> f b) -> f b
+--- helpers 
+(>>=) :: List a -> (a -> List b) -> List b
+(<*>) :: List (a -> b) -> List a -> List b
+pure :: a -> List a
+fmap :: (a -> b) -> List a -> List b
+replicate (1 :: Int) :: a -> [a]
+concatList :: List (List a) -> List a
+append :: List a -> List a -> List a
+
+I didnt understand which way it should behave
+>>> smartReplicate 3 (Cons 1 $ Cons 2 $ Empty)
+Cons [1,1,1] (Cons [2,2,2] Empty)
+>>> smartReplicate2 3 (Cons 1 $ Cons 2 $ Empty)
+Cons 1 (Cons 1 (Cons 1 (Cons 2 (Cons 2 (Cons 2 Empty))))
+Nevermind though, I implemented both anyway
```

Comment:
Yeah, both are correct, but to use the `Monad` instance, you need to have the same type for lists of lists. So it should be either `List (List Int)` or `[[Int]]`

The `Monad` instance for ordinary list implemented with the same meaning as for your `List` data type. If you have a list of values, and a function from an element to a list, then you can map all elements and get a list of lists, and then flatten the lists. 

If you look at the implementation of `smartReplicate` from Chapter Two, you can notice some similarities between `concatMap` and `>>=` for lists. And, indeed, `>>=` for lists is `concatMap`